### ISSUE TITLE
앨범 분류 기능 구현 모델 배포

### ISSUE THREAD

**Issue by mic006016** [B-Tripy/Tripy#22]
### 구현 기능 제목
- 딥러닝(CNN) 전이학습을 통한 이미지 카테고리 분류 모델 생성

### 주요 구현 내용 요약
- 데이터셋 구축
- 전이학습
- 학습 결과 분석
- Data Cleaning
- 모델 재학습
- 모델 저장

---

<html>
<body>

<h3>1. 데이터 준비 및 탐색</h3>
<ul>
<li>데이터 취합 : AI 허브, kaggle</li>
</ul>
<h3>2. 데이터 분할 및 전처리 (Preprocessing &amp; Augmentation)</h3>
<ul>
<li>리사이징, 텐서 변환</li>
<li><strong>작업</strong>:
<ul>
<li><strong>Split</strong>: Train / Validation / Test 세트로 분할 (8:1:1).</li>
<li><strong>Transform</strong>: Resize, Normalization.
→ 구글 드라이브 업로드 위한 <strong>오프라인 리사이징</strong>
<ul>
<li>리사이징: 224x224 크기로 축소 (물리적 픽셀 감소)</li>
<li>압축: JPEG 포맷으로 변환하고 품질을 약간 낮춰(85%) 용량 최적화</li>
</ul>
</li>
<li>Pycharm에서 데이터 전처리 후 Colab으로 이동</li>
</ul>
</li>
</ul>
<h3>3. 사전 학습 모델 로드 (Load Pre-trained Model)</h3>
<p>처음부터 가중치를 만드는 대신, 이미 거대한 데이터(ImageNet)로 학습된 모델 가져오기.</p>
<ul>
<li><strong>런타임 설정</strong>: GPU - A100 / 고용량 RAM</li>
<li><strong>모델</strong>: <code>EfficientNetV2-S</code></li>
<li><strong>Freeze</strong>: 앞쪽의 특징 추출기(Feature Extractor) 부분은 학습되지 않도록 고정하고, 마지막 출력층(Classifier)만 내 데이터의 클래스 개수(8개: buildings, culture 등)에 맞게 교체.</li>
<li>모델 GPU로 이동</li>
</ul>
<p><strong>“계산은 GPU(PyTorch)가 하고, 결과 분석과 그래프는 CPU(Scikit-learn, Matplotlib)가 수행”</strong></p>
<h3>4. 모델 설정 및 학습 (Training)</h3>
<p>실제로 GPU를 사용하여 가중치를 업데이트하는 단계.</p>
<ul>
<li>학습 전 모델 성능: 10.48% (랜덤 찍기 확률과 동일)</li>
<li>학습 후 모델 성능: <strong>94.86%</strong></li>
<li><strong>작업</strong>:
<ul>
<li>
<p><strong>손실 함수 (Loss Function):</strong> <code>nn.CrossEntropyLoss()</code></p>
<p>⇒ 클래스가 2개가 아니라 <strong>8개</strong>이기 때문에, 다중 분류에 필수적인 Cross Entropy를 사용했습니다.
(내부적으로 Softmax 확률 계산이 포함되어 있음.)</p>
</li>
<li>
<p><strong>최적화 기법 (Optimizer):</strong> <code>optim.Adam</code></p>
<ul>
<li>
<p><strong>학습 대상:</strong> <code>model.classifier.parameters()</code> (<strong>★중요</strong>)</p>
<p>⇒ 모델 전체(<code>model.parameters()</code>)를 학습시키는 게 아니라, 우리가 새로 교체한 <strong>마지막 분류기 층(Classifier)만</strong> 학습하도록 설정. (전이학습의 핵심)</p>
</li>
<li>
<p><strong>학습률 (Learning Rate):</strong> <code>0.001</code></p>
<ul>
<li>Adam에서 가장 무난하게 잘 작동하는 기본값.</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>Epochs</strong>: 10</p>
<ul>
<p>매 에포크(Epoch)마다 Validation 정확도 확인.</p>
</ul>
</li>

</li>
</ul>
</li>
</ul>
</li>
</ul>

<h3>5. 모델 평가 및 결과 분석 (Evaluation &amp; Inference)</h3>
<p>최종 성능을 확인하고 실제 이미지로 테스트</p>
- 최종 test 정확도: 94.86%

1. **Confusion Matrix**(어떤 클래스를 헷갈려 하는지) 시각화.

<img width="500" alt="Image" src="https://github.com/user-attachments/assets/dcd750b3-bbb9-4409-9bbb-5de44ba02df1" />

**(1) 빌딩(Buildings) vs 거리(Street) : 가장 큰 구멍**

- **현상:**
    - 실제 **`buildings`** 인데 **`street`** 라고 착각한 게 **30개**나 됨. 
    (전체 오답 중 압도적 1위)
    - 반대로 실제 **`street`** 인데 **`buildings`** 라고 착각한 건 **14개임.**
- **원인 분석:**
    - "거리" 사진에는 당연히 옆에 "건물"이 찍혀 있고, "건물" 사진도 아래에 "도로"가 깔려 있는 경우가 많음.
    - 모델 입장에서는 **"건물이 메인이냐, 도로가 메인이냐"**를 판단하기 어려운 것.
- **해결책:**
    - `show_worst_errors` 코드로 이 30장의 사진 검토하기. 라벨링이 애매하게 되어 있을 확률이 높음. (예: 건물이 꽉 찬 사진인데 라벨이 Street로 되어 있거나 그 반대)

**(2) 바다(Sea) vs 산(Mountain) : 자연 풍경의 혼동**

- **현상:**
    - **`sea`** 를 **`mountain`** 으로 착각 (9개)
    - **`mountain`** 을 **`sea`** 로 착각 (6개)
- **원인 분석:**
    - 멀리서 찍은 풍경 사진에서 **"파란 하늘과 수평선"** vs **"파란 하늘과 산등성이"**의 경계가 모호할 때가 있음.
    - 혹은 산 속에 호수가 있거나, 바다 뒤에 산이 있는 사진일 수 있음.

**(3) 문화(Culture)의 정체성**

- **현상:**
    - `culture`를 `buildings`(11개)나 `street`(8개)로 헷갈림.
- **원인 분석:**
    - `culture` 카테고리 안에 "궁궐, 절(건물)"이나 "퍼레이드(거리)" 사진이 포함되어 있을 것임.
    - 모델은 이걸 "문화재"라는 추상적 개념이 아니라, 그냥 시각적으로 비슷한 "건물"이나 "길"로 인식했을 가능성이 큼.
    

2. **Classification Report(F1-Score)**
    

<img width="381" height="263" alt="Image" src="https://github.com/user-attachments/assets/96e20b71-e246-4a4a-8285-faeee815b60a" />

**총점: A (94.86%)**
하지만 '우등생 과목'과 '열등생 과목'이 극명하게 나뉨.

✅ **우등생 과목 (완벽함)**

- **`food` (음식):** F1-score **0.9922**. 514장 중 오직 5장만 틀림.
- **`person` (사람):** F1-score **0.9760**. 사람이 포함된 이미지는 거의 완벽하게 구분.
- **`forest` (숲):** F1-score **0.9800**. 초록색 텍스처를 아주 잘 잡음.

⚠️ **열등생 과목 (집중 케어 필요)**

- **`buildings` (건물):**
    - **F1-score 0.8663** (전체 꼴찌)
    - **재현율(Recall) 0.83:** 실제 건물 사진 100장 중 17장을 놓치고 있음.
    - **오답 분석:** `street`(30장), `culture`(11장)로 가장 많이 샘.
- **`street` (거리):**
    - **F1-score 0.8919** (뒤에서 2등)
    - **오답 분석:** `buildings`(14장)로 주로 헷갈려함.

3. **Grad-CAM**

<img width="600" alt="Image" src="https://github.com/user-attachments/assets/a56ca8c8-2b36-406a-91ae-1f22b4f90c2a" />

4. **Worst Errors**
    - 총 오답 개수: 132개
    - 가장 높은 확신을 가지고 오답을 예상한 상위 6개

<img width="600" alt="Image" src="https://github.com/user-attachments/assets/d8a3dc8c-6eb3-4fac-b619-f44335f04069" />

- 가장 뻔뻔한 오답 + Grad-CAM 시각화

<img width="600" alt="Image" src="https://github.com/user-attachments/assets/6185247a-dca3-4750-9042-d32e097bdc09" />

---

### **'건물 vs 거리’ Grad-CAM 심층 분석 (오답의 원인)**

<img width="600" alt="Image" src="https://github.com/user-attachments/assets/e276ddd0-8232-4a0b-884a-826daf43c4e4" />

**패턴 A: "난 바닥(도로)만 보는데?" (The Floor Observer)**

- **현상:** 모델이 건물의 웅장한 윗부분은 안 보고, 사진 하단의 **'도로', '자동차', '인도'**에만 빨간불(집중)을 켜고 있음.
- **해석:** EfficientNet에게 "자동차나 아스팔트가 보이면 무조건 Street야!"라는 편견이 생긴 것 같아보임.

**패턴 B: "상점 간판은 거리의 상징이지" (Signboard Effect)**

- **현상:** 건물 전체보다는 **1층의 상점 간판, 쇼윈도**에 집중함.
- **해석:** 학습 데이터 중 `street` 클래스에 '상점가 거리' 사진이 많았던 것으로 추정됨. 그래서 "간판 = 거리"라는 공식이 성립된 듯함.

**패턴 C: "솔직히 이건 사람도 헷갈린다" (Ambiguous Data)**

- **현상:** 사진 구도 자체가 **'건물을 찍은 건지, 거리를 찍은 건지'** 모호함.
- 건물이 화면에 꽉 차기보다, 소실점이 있는 '길'의 구도로 찍힌 사진이 많음.

<h3>6. Data Cleaning</h3>

`train/buildings` 폴더 안에 숨어있는 "거리(Street) 같은 녀석들"을 자동으로 색출<br>

1. **전수 조사:** train, valid, test 별 buildings 사진 검사.<br>
2. **색출:** 모델이 "어? 이건 `buildings`가 아니라 `street` 같은데?"라고 판단한 사진 골라내기.<br>
3. **격리:** 해당 사진들을 지우기 전에, **`/content/cleaning_candidates`** 라는 별도 폴더로 이동. (안전을 위해 바로 삭제하지 않고 격리부터 수행.)<br>

  1) [구조 작업] 억울한 파일 복구하기<<br>
  2) 격리 폴더에 남은 이미지 street 폴더로 이사<br>
  3) valid, test 사진들도 동일한 방식으로 수행<br>

**[ buildings → street 라벨링 수정 ]**
- train : 36/42 (confidence 0.9 이상)
- valid : 8/11 (confidence 0.9 이상)
- test : 12/31 (buildings → street 오답)

<h3>7. 모델 재학습 분석</h3>

성능 변화: 94.86% -> **96.02% (+1.16%p)**
데이터 청소 효과로 성능이 향상됨.

- **데이터셋 갱신:** 파일 이동/삭제 후 `ImageFolder`와 `DataLoader`를 다시 선언하여 변경 사항을 반영한 점.
- **최고 모델 저장 (`deepcopy`):** 앞서 설명드린 대로 `best_model_wts2`에 가중치를 깊은 복사로 따로 저장하여, 학습 후반부에 성능이 떨어지는 것을 방지한 점.
- **검증 모드 분리:** `train`과 `eval` 모드를 구분하고, 검증(`valid`) 단계에서는 기울기 계산(`grad`)을 꺼서 메모리를 아낀 점.

---

1. **Confusion Matrix**

<img width="500" alt="Image" src="https://github.com/user-attachments/assets/007f64b1-ecee-4b33-9c1f-a01c8f66d82b" />

**<특정 클래스 간의 혼동 문제**를 명확하게 해결하려 노력>

**1. 가장 큰 개선점: 'Buildings'와 'Street'의 구분 능력 향상**

첫 번째 모델(EfficientNetV2-S)에서 가장 큰 문제는 **'Buildings(건물)'를 'Street(거리)'로 착각하는 것**이었음. 재학습 모델에서는 이 부분이 눈에 띄게 개선됨.

- **Buildings(건물) 클래스:**
    - **처음:** 실제 건물 사진 30장을 '거리'로 잘못 예측함. 
    (오답률이 가장 높은 구간)
    - **재학습 후:** '거리'로 잘못 예측한 개수가 **11장으로 대폭 감소함**. 정답(True Positive) 개수도 217개에서 229개로 증가함.
- **원인 추정 (데이터 정제 가능성):**
    - 흥미로운 점은 `True Label(Actual)`의 총합이 변경됨.
    - 처음 모델: Buildings 행의 합(실제 데이터 수)은 **262개**, Street는 **288개**.
    - 재학습 모델: Buildings 행의 합은 **250개**, Street는 **300개**.
    - **분석:** 정확히 **12개의 이미지가 Buildings에서 Street로 라벨이 변경**된 것으로 보임. 즉, 모델 파라미터만 조정한 것이 아니라, **잘못 라벨링 된 데이터를 수정(Data Cleaning)하고 학습**시켰기 때문에 성능이 비약적으로 상승했을 가능성이 매우 높음. 좋은 접근.

**2. 견고한 클래스들의 성능 유지 및 향상**

이미 잘 맞추던 클래스들은 성능이 유지되거나 소폭 상승함.

- **Food(음식):** 509개 → 513개로 정답 수가 늘었음. 오분류가 거의 없는 가장 완벽한 클래스.
- **Sea(바다):** 260개 → 265개로 정답 수가 증가함.
- **Culture, Forest:** 여전히 높은 정확도를 유지하고 있음.

**3. 여전히 남아있는 미세한 혼동 (Minor Issues)**

큰 문제는 아니지만, 모델이 헷갈려 하는 패턴이 소수 존재함.

- **Sea(바다) vs Mountain(산):**
    - 배경이 비슷해서인지 서로를 착각하는 경우가 여전히 있음 (재학습 모델 기준: 바다를 산으로 7장, 산을 바다로 8장 오분류).
    - 이는 풍경 사진 특성상 수평선이나 산등성이의 라인이 비슷하기 때문일 수 있음.
- **Street(거리) → Buildings(건물) 오분류:**
    - '건물'을 '거리'로 보는 실수는 줄었지만, 반대로 **'거리' 사진을 '건물'로 예측하는 실수**는 14건에서 18건으로 소폭 증가함. 거리 사진 내에 건물이 크게 찍힌 데이터들이 일부 있을 것으로 추정됨.

### 요약 및 제언

데이터 라벨링을 수정한 것(Buildings → Street)이 '**신의 한 수'**. 모델이 건물과 거리의 특징을 훨씬 더 명확하게 학습하게 됨.

---

2. **Classification Report(F1-Score)**

<img width="382" height="266" alt="Image" src="https://github.com/user-attachments/assets/1a012569-3767-44cb-9b2a-f5807c092626" />

**1. 핵심 성과: 'Buildings'와 'Street'의 드라마틱한 개선**

이 두 클래스가 전체 성능 향상을 주도함.

- **Buildings(건물): Recall(재현율)의 급상승 (0.8282 → 0.9160)**
    - **의미:** 이전 모델은 실제 건물을 보고도 건물이라고 인식하지 못하는 비율이 높았음(Recall 낮음).
    - **변화:** 재학습 후, **실제 건물을 건물로 정확히 찾아내는 능력**이 대폭 향상됨. F1-Score가 0.83에서 **0.90**대로 진입한 것은 상징적인 의미가 큼.
- **Street(거리): Precision(정밀도)의 급상승 (0.8684 → 0.9410)**
    - **의미:** 이전 모델은 "이건 거리야"라고 예측했을 때, 실제로는 거리가 아닌 경우(주로 건물)가 많았음.
    - **변화:** 재학습 후, 모델이 **"거리"라고 예측했을 때의 신뢰도**가 매우 높아짐.
- **원인 (Support 수치 변화):**
    - 표의 맨 오른쪽 `support` 열을 보면, **Buildings는 262→250**으로 줄고, **Street는 288→300**으로 늘었음.
    - 이는 앞서 추론한 대로 **잘못 라벨링 된 데이터(건물로 분류된 거리 사진)를 올바른 클래스(Street)로 이동**시켰음을 확실히 보여줌. 이 데이터 정제 작업이 두 클래스 모두의 성능을 끌어올림.

**2. Sea(바다)와 Mountain(산)의 트레이드오프**

- **Sea (바다):** Recall은 올랐으나(0.93→0.95), **Precision은 소폭 하락(0.95→0.93)**함.
    - 이는 바다를 산으로, 혹은 산을 바다로 헷갈리는 경향이 조금 남아있거나, 모델이 바다를 좀 더 과감하게 예측하면서 생긴 현상일 수 있음. 하지만 전체적인 F1-Score는 소폭 상승(0.9420 → 0.9447)하여 긍정적임.
- **Mountain (산):** Precision이 소폭 상승함(0.96→0.97). 산이라고 예측하면 거의 다 산이라는 뜻임.

**3. 'Food'와 'Person'은 이미 완성형**

- **Food:** F1-Score **0.9952**. 더 이상 손댈 곳이 없을 정도로 완벽함.
- **Person:** F1-Score **0.9909**. 역시 매우 훌륭함.

**4. 종합 평가**

- **전체 정확도(Accuracy): 94.86% → 96.02%**
- **균형 잡힌 성능 (Macro Avg 상승):**
    - 단순히 쉬운 클래스만 잘 맞춘 것이 아니라, 모든 클래스의 평균 성능(Macro Avg)이 **0.9447 → 0.9580**으로 1.3% 가까이 올랐음.

### **결론**

데이터 라벨링 수정 전략이 **정밀도(Street)**와 **재현율(Buildings)** 두 마리 토끼를 모두 잡는 결과를 낳음. 현재 상태는 배포(Deployment)나 실제 프로젝트에 적용하기에 손색없는 매우 우수한 수준임.

<h3>8. 모델 저장</h3>
모델 배포 ⇒ efficientnet_v2_final.pth

코드랩 계정 구글드라이브
https://drive.google.com/file/d/1uiZcBSko7y47VdRTUD4YVQOdhpkb4V10/view?usp=drive_link

</body>
</html>
